pkg/instance/backupService.go

```diff
@@ -0,0 +1,240 @@
+	s3Client    S3BackupClient
+	logger      *slog.Logger
+}
+
+// MinioClient defines the methods we need from MinIO client
+type MinioClient interface {
+	ListObjects(ctx context.Context, bucketName string, opts minio.ListObjectsOptions) <-chan minio.ObjectInfo
+	GetObject(ctx context.Context, bucketName, objectName string, opts minio.GetObjectOptions) (*minio.Object, error)
+}
+
+// S3BackupClient defines the methods we need from AWS S3 client
+type S3BackupClient interface {
+	CreateMultipartUpload(ctx context.Context, params *s3.CreateMultipartUploadInput, optFns ...func(*s3.Options)) (*s3.CreateMultipartUploadOutput, error)
+	UploadPart(ctx context.Context, params *s3.UploadPartInput, optFns ...func(*s3.Options)) (*s3.UploadPartOutput, error)
+	CompleteMultipartUpload(ctx context.Context, params *s3.CompleteMultipartUploadInput, optFns ...func(*s3.Options)) (*s3.CompleteMultipartUploadOutput, error)
+	AbortMultipartUpload(ctx context.Context, params *s3.AbortMultipartUploadInput, optFns ...func(*s3.Options)) (*s3.AbortMultipartUploadOutput, error)
+}
+
+// BackupStats tracks backup operation statistics
+type BackupStats struct {
+	ObjectsProcessed int64
+	BytesProcessed   int64
+	StartTime        time.Time
+	mu               sync.Mutex
+}
+
+func (bs *BackupStats) increment(size int64) {
+	bs.mu.Lock()
+	defer bs.mu.Unlock()
+	bs.ObjectsProcessed++
+	bs.BytesProcessed += size
+}
+
+// PerformBackup executes the backup operation
+func (s *BackupService) PerformBackup(ctx context.Context, minioBucket, s3Bucket, key string) error {
+	g, ctx := errgroup.WithContext(ctx)
+	stats := &BackupStats{StartTime: time.Now()}
+	pr, pw := io.Pipe()
+
+	g.Go(func() error {
+		defer pw.Close()
+		return s.createTarGzStream(ctx, pw, minioBucket, stats)
+	})
+
+	g.Go(func() error {
+		return s.streamToS3WithMultipart(ctx, s3Bucket, key, pr)
+	})
+
+	if err := g.Wait(); err != nil {
+		return fmt.Errorf("backup failed: %v", err)
+	}
+
+	s.logBackupStats(stats)
+	return nil
+}
+func (s *BackupService) createTarGzStream(ctx context.Context, w io.Writer, minioBucket string, stats *BackupStats) error {
+	gw := gzip.NewWriter(w)
+	defer gw.Close()
+
+	tw := tar.NewWriter(gw)
+	defer tw.Close()
+
+	objectCh := s.minioClient.ListObjects(ctx, minioBucket, minio.ListObjectsOptions{
+		Recursive: true,
+	})
+
+	for object := range objectCh {
+		if object.Err != nil {
+			return fmt.Errorf("list objects: %v", object.Err)
+		}
+
+		if err := s.processSingleObject(ctx, tw, minioBucket, object, stats); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *BackupService) processSingleObject(ctx context.Context, tw *tar.Writer, minioBucket string, object minio.ObjectInfo, stats *BackupStats) error {
+	s.logger.InfoContext(ctx, "Processing object", "key", object.Key, "bucket", minioBucket)
+	obj, err := s.minioClient.GetObject(ctx, minioBucket, object.Key, minio.GetObjectOptions{})
+	if err != nil {
+		return fmt.Errorf("get object %s: %v", object.Key, err)
```

Comment:
Shouldn't this and the errors below start with "failed to"? How does the error look like to the user on aborting an upload?